### PR TITLE
Implement RefineNet

### PR DESCRIPTION
**Pull request by jiayukeyme**
This commit implements RefineNet in [1] Guosheng Lin, Anton Milan, Chunhua Shen, Ian Reid RefineNet: Multi-Path Refinement Networks for High-Resolution Semantic Segmentatiion. arXiv:1611.06612.

Main function in `refinenet.py` is used to check the model can build and run.

Since RefineNet relies on ResNet, `resnet_v2.py` and `resnet_utils.py` from `tensorflow.models.research.slim` are also added to this repo.

### CODE REVIEW

File: refinenet/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+
+* RefineNet was originally proposed in:
+[1] Guosheng Lin, Anton Milan, Chunhua Shen, Ian Reid
+    RefineNet: Multi-Path Refinement Networks for High-Resolution Semantic Segmentatiion. arXiv:1611.06612
+
+* The RefineNet implemented in this module is based on [1], 
+along with ResNet-50 (resnet_v2 from slim is used) 
+
+* There key differences between this implementation and the one proposed in [1]:
```
Comment (kborer) [inline refinenet/README.md]: The key

File: refinenet/README.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    1. [1] uses pretrained ResNet while this implementaiton is an end-to-end version
```
Comment (kborer) [inline refinenet/README.md]: implementation

File: refinenet/refinenet.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+The RefineNet implemented in this module is based on [1], 
+along with ResNet-50 (resnet_v2 from slim is used) 
+
+There key differences between this implementation and the one proposed in [1]:
+1. [1] uses pretrained ResNet while this implementaiton is an end-to-end version
+2. [1] uses 512 for RefineNet-4 block, while this implementation uses 256 instead, 
+   to keep it consistent with the remaining RefineNet blocks
+
+"""
```
Comment (kborer) [inline refinenet/refinenet.py]: probably don't need this comment block in both files

File: refinenet/refinenet.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        axis_to_reduce = (0, 1, 2) if self.data_format == "channels_last" else (0, 2, 3)
+        mean_pixel = tf.reduce_mean(self.placeholders["image"], axis=axis_to_reduce)
+        processed_image = self.placeholders["image"] - mean_pixel
+
+        if self.data_format == "channels_first":
+            processed_image = tf.transpose(processed_image, [0, 3, 1, 2])
+
+        num_filters = NUM_REFINE_FILTERS
+
+        def rcu(x):
```
Comment (kborer) [inline refinenet/refinenet.py]: Many of the helper functions here are named with acronyms. The code would be easier to read if you used longer names and would remove the need to add comments that explain what the function calls do.

File: refinenet/refinenet.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+                low_res_adap_conv_out = rcu(rcu(low_res_input))
+            high_res_adap_conv_out = rcu(rcu(high_res_input))
+
+            # Multi-resolution Fusion
+            mrf_out = mrf(low_res_adap_conv_out, high_res_adap_conv_out)
+
+            # Chained Residual Pooling
+            crp_out = crp(mrf_out)
+
+            # Output Conv
```
Comment (kborer) [inline refinenet/refinenet.py]: this comment isn't adding much

File: refinenet/refinenet.py [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+                          end_points['resnet_v2/block2/unit_3/bottleneck_v2'],
+                          end_points['resnet_v2/block3/unit_5/bottleneck_v2'],
+                          end_points['resnet_v2/block4']]
+
+        # refine outputs recursively
+        refine_out = None
+        refine_out = refine_block(refine_out, multi_res_outs[3])
+        refine_out = refine_block(refine_out, multi_res_outs[2])
+        refine_out = refine_block(refine_out, multi_res_outs[1])
+        refine_out = refine_block(refine_out, multi_res_outs[0])
```
Comment (kborer) [inline refinenet/refinenet.py]:
```python
for i in [3,2,1,0]:
    refine_out = refine_block(refine_out, multi_res_outs[i])

File: refinenet/refinenet.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+
+        # two more rcus right before the final scoring layer
+        refine_out = rcu(rcu(refine_out))
```
Comment (kborer) [inline refinenet/refinenet.py]: slight preference for a loop here

File: refinenet/refinenet.py [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+        if self.data_format == "channels_first":
+            logits = tf.transpose(logits, [0, 2, 3, 1])
+
+        # bilinearly up-sample to match the dimension of the input image
+        logits = tf.image.resize_images(logits, self.input_shape)
+
+        annotation_pred = tf.argmax(logits, axis=3, name="prediction", output_type=tf.int32)
+
+
```
Comment (kborer) [inline refinenet/refinenet.py]: rm

**Review by kborer — COMMENTED** [review]
Some minor things. Please also mention in the README where you copied the resnet files from.

**Comment by jiayukeyme** [comment]
Added the reference for `resnet_v2.py` and `resnet_utils.py` in `README.md`.
Fixed naming and style issues.

File: refinenet/README.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    RefineNet: Multi-Path Refinement Networks for High-Resolution Semantic Segmentatiion. arXiv:1611.06612
+
+* The RefineNet implemented in this module is based on [1]
+
+* Since RefineNet relies on ResNet, we use ResNet-50 for this implementation
+    * `resnet_v2.py` and `resnet_utils.py` are copied from github repo `tensorflow/models/research/slim/nets`
+    * resnet_v2_50 in `resnet_v2.py` is used to create the 50-layer ResNet
+
+* The key differences between this implementation and the one proposed in [1]:
+    * [1] uses pretrained ResNet while this implementaton is an end-to-end version
```
Comment (nickmarton) [inline refinenet/README.md]: Can you change this to include saying which dataset it was pretrained on; also `while this implementation is trained end-to-end` is preferable

File: refinenet/README.md [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    * [1] uses 512 filters for each conv layer of RefineNet-4 block, while this implementation uses 256 instead, 
+   to keep it consistent with the remaining RefineNet blocks
```
Comment (nickmarton) [inline refinenet/README.md]: why is this on a separate line?

File: refinenet/refinenet.py [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+NUM_REFINE_FILTERS = 256
+INPUT_IMAGE_SHAPE = (512, 512)
+
+class RefineNet(object):
+    def __init__(self,
+                 input_shape,
+                 is_training,
+                 num_classes,
+                 learning_rate=1e-4,
+                 network_storage_dir=None):
```
Comment (nickmarton) [inline refinenet/refinenet.py:18]: Add a docstring; even if it's something simple like `"""Construct the RefineNet object."""`

File: refinenet/refinenet.py [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+
+        self.placeholders = {"image": image,
+                             "annotation": annotation}
+
+        # Build the computational graph
+        self.predict_with_network, self.logits = self.build_computational_graph()
+
+        self.loss = tf.reduce_mean(
+                       (tf.nn.sparse_softmax_cross_entropy_with_logits(
+                        logits=self.logits, labels=tf.squeeze(annotation, squeeze_dims=[3]), name="entropy")))
```
Comment (nickmarton) [inline refinenet/refinenet.py]: can you rename to `xentropy`

File: refinenet/refinenet.py [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+
+        # get train op
+        trainable_var = tf.trainable_variables()
+        self.train_op = self.train(self.loss, trainable_var)
+
+        self.num_params = self.compute_total_num_params()
+
+        self.sess = tf.Session()
+
+        # Set up the saver and try to restore any model previously being trained
```
Comment (nickmarton) [inline refinenet/refinenet.py]: Remove this comment considering the comment underneath

File: refinenet/refinenet.py [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+                                    data_format=self.data_format)
+
+            # residual connection
+            compare_dimension = 1 if self.data_format == "channels_first" else 3
+            if x.get_shape()[compare_dimension] == conv.get_shape()[compare_dimension]:
+                # if same dimension, then add them up
+                out = tf.add(x, conv)
+            else:
+                # if different dimension, then project shortcut x to have the same dimension with conv
+                x_conv = tf.layers.conv2d(x, filters=num_filters,
```
Comment (nickmarton) [inline refinenet/refinenet.py]: Can you expand on this comment a little bit for less informed readers? I.e. can you mention that since padding is the same the only dimension along with `x` and `conv` can differ is channel depth hence the 1x1 convolution to correct to right num channels?

File: refinenet/refinenet.py [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+        axis_to_reduce = (0, 1, 2) if self.data_format == "channels_last" else (0, 2, 3)
+        mean_pixel = tf.reduce_mean(self.placeholders["image"], axis=axis_to_reduce)
+        processed_image = self.placeholders["image"] - mean_pixel
+
+        if self.data_format == "channels_first":
+            processed_image = tf.transpose(processed_image, [0, 3, 1, 2])
+
+        num_filters = NUM_REFINE_FILTERS
+
+        def residual_conv_unit(x):
```
Comment (nickmarton) [inline refinenet/refinenet.py:92]: Add a docstring here

File: refinenet/refinenet.py [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+                out = tf.add(x, conv)
+            else:
+                # if different dimension, then project shortcut x to have the same dimension with conv
+                x_conv = tf.layers.conv2d(x, filters=num_filters,
+                                          kernel_size=1, strides=1, padding="same",
+                                          data_format=self.data_format)
+                out = tf.add(x_conv, conv)
+            return out
+
+        def multi_resolution_fusion(low_res_input, high_res_input):
```
Comment (nickmarton) [inline refinenet/refinenet.py:118]: docstring

File: refinenet/refinenet.py [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+            def conv_and_upsample(x, stride):
```
Comment (nickmarton) [inline refinenet/refinenet.py]: docstring mentioning this is a helper function

File: refinenet/refinenet.py [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+                conv = tf.layers.conv2d(x, filters=num_filters, kernel_size=3,
+                                        strides=stride, padding="same",
+                                        data_format=self.data_format)
+                # transposed convolutional layer to upsample the feature map
+                upsample = tf.layers.conv2d_transpose(conv, filters=num_filters, kernel_size=3,
+                                                      strides=2, padding="same",
+                                                      data_format=self.data_format)
+                return upsample
+            # we assume both the width and height of low_res_out are 1/2 of high_res_out
```
Comment (nickmarton) [inline refinenet/refinenet.py]: Can you move this comment into the docstring and also specify that our MRF function only handles exactly 2 inputs (diverges slightly from paper here which is why it's worth mentioning)

File: refinenet/refinenet.py [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+            if low_res_input is not None:
+                # we upsample the low resolution input to have the same dimension of high resolution input
+                low_res_out = conv_and_upsample(low_res_input, stride=1)
+            high_res_out = conv_and_upsample(high_res_input, stride=2)
+            if low_res_input is not None:
+                return tf.add(low_res_out, high_res_out)
+            else:
+                return high_res_out
+
+        def chained_residual_pooling(x):
```
Comment (nickmarton) [inline refinenet/refinenet.py:141]: docstring

File: refinenet/refinenet.py [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+            def pool_and_conv(x):
```
Comment (nickmarton) [inline refinenet/refinenet.py]: docstring saying this is a helper func

File: refinenet/refinenet.py [hunk 21]
```diff
@@ -0,0 +1,253 @@
...
+                ksize = [1, 1, 5, 5] if self.data_format == "channels_first" else [1, 5, 5, 1]
+                data_format = "NCHW" if self.data_format == "channels_first" else "NHWC"
+                max_pool = tf.nn.max_pool(x, ksize=ksize, strides=[1, 1, 1, 1], padding="SAME", data_format=data_format)
+                conv = tf.layers.conv2d(max_pool, filters=num_filters,
+                                        kernel_size=3, strides=1, padding="same",
+                                        data_format=self.data_format)
+                return conv
```
Comment (nickmarton) [inline refinenet/refinenet.py:151]: Add a newline after this line

File: refinenet/refinenet.py [hunk 22]
```diff
@@ -0,0 +1,253 @@
...
+            relu = tf.nn.relu(x)
+            pool_conv_out = relu
+            sum_out = relu
+            num_pool_conv_units = 2
```
Comment (nickmarton) [inline refinenet/refinenet.py]: Can you add a comment saying this was a design choice and one can easily extend the CRP by increasing this value

File: refinenet/refinenet.py [hunk 23]
```diff
@@ -0,0 +1,253 @@
...
+            for i in range(num_pool_conv_units):
```
Comment (nickmarton) [inline refinenet/refinenet.py]: Nothing to fix here; just remarking that this is a clever way of renaming the tensors here to keep things concise

File: refinenet/refinenet.py [hunk 24]
```diff
@@ -0,0 +1,253 @@
...
+
+        # get outputs from ResNet that we want to refine on
+        multi_res_outs = [end_points['resnet_v2/block1/unit_2/bottleneck_v2'],
+                          end_points['resnet_v2/block2/unit_3/bottleneck_v2'],
+                          end_points['resnet_v2/block3/unit_5/bottleneck_v2'],
+                          end_points['resnet_v2/block4']]
+
+        # refine outputs recursively
+        refine_out = None
+        for i in [3, 2, 1, 0]:
```
Comment (nickmarton) [inline refinenet/refinenet.py]: Can you add a comment about the values in this list?

File: refinenet/refinenet.py [hunk 25]
```diff
@@ -0,0 +1,253 @@
...
+        logits = tf.image.resize_images(logits, self.input_shape)
+
+        annotation_pred = tf.argmax(logits, axis=3, name="prediction", output_type=tf.int32)
+
+        return tf.expand_dims(annotation_pred, dim=3), logits
+
+    def train(self, loss_val, var_list):
+        """Add training operations to computational graph."""
+        self.global_step = tf.get_variable(name="global_step", dtype=tf.int32, initializer=0)
+        update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
```
Comment (nickmarton) [inline refinenet/refinenet.py:229]: Why this over `trainable_variables()`?

File: refinenet/refinenet.py [hunk 26]
```diff
@@ -0,0 +1,253 @@
...
+        with tf.control_dependencies(update_ops):
+            optimizer = tf.train.AdamOptimizer(self.learning_rate)
+            grads = optimizer.compute_gradients(loss_val, var_list=var_list)
+            return optimizer.apply_gradients(grads, self.global_step)
+
+if __name__ == "__main__":
+    """ run this main function to test the model """
```
Comment (nickmarton) [inline refinenet/refinenet.py:236]: The default main procedure should be training/testing on a real dataset; the contents of `__main__` right now would be better off as a separate test function. Is this here because of difficulty in getting the paper's datasets?

**Review by nickmarton — CHANGES_REQUESTED** [review]
Minor stylistic stuff and a few questions, but lgtm otherwise

File: refinenet/refinenet.py [hunk 27]
```diff
@@ -0,0 +1,253 @@
...
+        # two more rcus right before the final scoring layer
+        for i in range(2):        
+            refine_out = residual_conv_unit(refine_out)
+        
+        # scoring layer
+        logits = tf.layers.conv2d(refine_out, filters=self.num_classes,
+                                      kernel_size=3, strides=1, padding="same",
+                                      data_format=self.data_format)
+
+        if self.data_format == "channels_first":
```
[the remark this answers is not in the document]
Comment (jiayukeyme) [reply refinenet/refinenet.py]: yes, just to make the output NHWC

File: refinenet/refinenet.py (hunk 18 above)
Comment (jiayukeyme) [reply refinenet/refinenet.py]: RefineNet with ResNet (50, 101, or 152) only fuses two inputs. I think Figure 3 in the paper shows a more generic version of MRF.

File: refinenet/refinenet.py (hunk 26 above)
Comment (jiayukeyme) [reply refinenet/refinenet.py:236]: The main function is solely to make sure the model can build and run.

PASCAL VOC is not hard to get, but ImageNet is a little bit. Need more time to integrate data preprocessing and training/testing functionalities to this module.

File: refinenet/refinenet.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (nickmarton) [reply refinenet/refinenet.py]: cool

File: refinenet/refinenet.py (hunk 26 above)
Comment (nickmarton) [reply refinenet/refinenet.py:236]: Sounds good, this is good for now, but we do need to add a public dataset so others can verify

File: refinenet/refinenet.py (hunk 25 above)
Comment (jiayukeyme) [reply refinenet/refinenet.py:229]: Mean and Std in batch normalization layer need to be updated during training. They need to be added as control dependencies explicitly.

Comment (nickmarton) [reply refinenet/refinenet.py:229]: Ahhhhh lol forgot that the resnet provided by tensorflow uses batch norm

File: refinenet/refinenet.py (hunk 26 above)
Comment (jiayukeyme) [reply refinenet/refinenet.py:236]: Yes, I'll work on this.

**Review by nickmarton — APPROVED** [review]
lgtm

**Comment by jiayukeyme** [comment]
Add a script `pretrain_resnet.py` to pre-train a ResNet-50 on a classification task. RefineNet can benefit from a pre-trained ResNet component.

The training is done on [Tiny ImageNet dataset (200 classes)](https://tiny-imagenet.herokuapp.com/), which is a subset of ImageNet dataset (1000 classes). A data wrangle script is also added to prepare Tiny ImageNet dataset to be used by Keras image generator.
